java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp

```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
+<qhelp>
+
+  <overview>
+    <p>
+      Android Contexts that depend on external inputs are potentially vulnerable to code execution attacks.          
+      This query detects the scenarios where an attacker controlled input is used to create a new Android <code> Context</code>. 
+    </p>
+  </overview>
+
+  <recommendation>
+    This can be avoided in two ways. 
+    <ol>
+      <li> 
+        To avoid arbitrary code execution, the application should sanitize all external inputs.
+      </li>
+      <li> Changing the application architecture to prevent unintended access to sensitive components.</li>
+
+
+  </recommendation>
+
+  <example>
+    <p>The following example shows a 'bad' way of creating Android `Context` variables. An attacker may register a package with the name `somename.plugin.` having a `pluginLoader`  class.</p>
+    <sample src="InsecureClassLoading.java" />
+    <p> Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.</p>
+  </example>
+
+  <references>
+    <li>
+      <a href="https://cwe.mitre.org/data/definitions/749.html">CWE-749</a>
+    </li>    
+  </references>
```

Comment:
If I recall correctly CWE links are  extracted from the `@tags` property on the query and automatically included when rendering a QHelp file . In addition you link to CWE-749 in this file but the query `@tags` mentions CWE-691 instead

```suggestion
```

Reply:
For that to work I think there needs to be a reference section - possibly with at least one other reference in it (which is good practice anyway).

Reply:
Ah curious, didn't know that. It seems to work fine with an empty references section. 

@porcupineyhairs You can "render" a qhelp file using the following command:
```
codeql generate query-help --format=markdown src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp
```

Reply:
Below is a rendered version of the QHelp after applying the suggestions I posted above:

<details>
<summary>Preview <code>InsecureClassLoading.qhelp</code></summary>

# Android insecure context creation
Android Contexts that depend on external inputs are potentially vulnerable to code execution attacks. This query detects the scenarios where an attacker controlled input is used to create a new Android ` Context`.


## Recommendation
This can be avoided in two ways.

1. To avoid arbitrary code execution, the application should sanitize all external inputs.
1. Changing the application architecture to prevent unintended access to sensitive components.

## Example
The following example shows a 'bad' way of creating Android \`Context\` variables. An attacker may register a package with the name \`somename.plugin.\` having a \`pluginLoader\` class.


```java

import android.app.Application;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.os.Bundle;

public class InsecureClassLoading extends Application {
  @Override
  public void onCreate() {
    super.onCreate();
    invokePlugins();
  }

  private void invokePlugins() {
    for (PackageInfo info : getPackageManager().getInstalledPackages(0)) {
      String packageName = info.packageName;
      Bundle meta = info.applicationInfo.metaData;
      if (packageName.startsWith("somename.plugin.")) {
        try {
          Context packageContext =
              createPackageContext(packageName, CONTEXT_INCLUDE_CODE | CONTEXT_IGNORE_SECURITY);
          packageContext.getClassLoader().loadClass("somename.pluginLoader")
              .getMethod("load", Context.class).invoke(null, this);
        } catch (Exception e) {
          throw new RuntimeException(e);
        }
      }
    }
  }
}

```
Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.


## References
* Common Weakness Enumeration: [CWE-691](https://cwe.mitre.org/data/definitions/691.html).

</details>

Reply:
The query rendering fails for me with a null pointer exception. Can you please take a look?
```
Oops! A fatal internal error occurred.
java.lang.NullPointerException
        at com.semmle.prettyprint.doc.QhelpToMD.checkFile(QhelpToMD.java:692)
        at com.semmle.prettyprint.doc.QhelpToMD.findFile(QhelpToMD.java:675)
        at com.semmle.prettyprint.doc.QhelpToMD.access$200(QhelpToMD.java:83)
        at com.semmle.prettyprint.doc.QhelpToMD$QhelpSaxHandler.startElement(QhelpToMD.java:389)
        at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)
        at org.apache.xerces.parsers.AbstractXMLDocumentParser.emptyElement(Unknown Source)
        at org.apache.xerces.impl.dtd.XMLDTDValidator.emptyElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanStartElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.SAXParserImpl.parse(Unknown Source)
        at com.semmle.prettyprint.doc.QhelpToMD.runSaxParse(QhelpToMD.java:814)
        at com.semmle.prettyprint.doc.QhelpToMD.process(QhelpToMD.java:764)
        at com.semmle.prettyprint.doc.QhelpToMD.format(QhelpToMD.java:627)
        at com.semmle.cli2.generate.QueryHelpCommand.executeSubcommand(QueryHelpCommand.java:134)
        at com.semmle.cli2.picocli.SubcommandCommon.call(SubcommandCommon.java:430)
        at com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:201)
        at com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:209)
        at com.semmle.cli2.CodeQL.main(CodeQL.java:91)
```

Reply:
There is a bug in file path handling. A call to `Path.getParent` returns `null` for a path like `InsecureClassLoading.qhelp` causing a NullPointerException in `.QhelpToMD.checkFile`. Could you try calling the command with `./InsecureClassLoading.qhelp` as a workaround?

I'll look into fixing the bug properly. Thanks for reporting.